js/apps/admin-ui/src/util.ts

```diff
@@ -171,3 +171,17 @@ export const addTrailingSlash = (url: string) =>
   url.endsWith("/") ? url : url + "/";
 
 export const generateId = () => Math.floor(Math.random() * 1000);
+
+export function generatePath<Path extends string>(
+  originalPath: Path,
+  params?: {
+    [key: string]: string | null;
+  },
+): string {
+  if (params)
+    Object.entries(params).forEach(
+      ([k, v]) => (params[k] = encodeURIComponent(v || "")),
+    );
+  //@ts-ignore as PathParam isn't an exported type
+  return generateUnencodedPath(originalPath, params);
+}
```

Comment:
A couple of points here:

1. Using the same name for the function as the library might cause future devs to accidentally import the one from the library, as the name is the same the IDE might autocomplete it, and it's easy to miss in review.
2. In our case we always pass in `params`, so there is no need for it to be optional.
3. Although React Router uses a value of `string | null` we always pass in `string`, so we can omit the `null` fallback.
4. Assigning `params[k]` in-place will mutate the original value value passed into the form, which might cause unwanted side effects. It would be better to create a copy and operate on that instead.
5. We can avoid using a `@ts-ignore` here, as we only need to cast `params` to `any`.

Putting that all together I think we'd be left with something like:

```ts
function generateEscapedPath<Path extends string>(
  originalPath: Path,
  params: Record<string, string>,
): string {
  const encodedParams = { ...params };

  for (const [key, value] of Object.entries(encodedParams)) {
    encodedParams[key] = encodeURIComponent(value);
  }

  // TODO: Fix type annotation once https://github.com/remix-run/react-router/pull/10719 is merged.
  return generatePath<Path>(originalPath, encodedParams as any);
}
```

Reply:
no need to put the code here, I think I know how to code